### ISSUE TITLE
OSMCha shows features and tags not edited in the selected changeset

### ISSUE THREAD

**Issue by musuruan** [mapbox/osmcha-frontend#604]
## *Brief Description*
OSMCha shows features and tags not edited in the selected changeset.

## What  is the current behaviour?
For example, in changeset #114816402, only one feature has been edited but OSMCha shows two:

![immagine](https://user-images.githubusercontent.com/1221995/146364197-eb82b433-857b-4f46-adb3-04b5d584dc12.png)

The other feature was edited in changeset #114816420.

Moreover, for the feature edited in the changeset OSMCha reports two tag changes, but only one was added. The other was added in changeset #114816412.

![immagine](https://user-images.githubusercontent.com/1221995/146364473-15ea8f4e-f1a7-4074-9443-62edc6b5311d.png)

## What  is the expected behaviour ?
Only changes in the current changeset should be shown.

## When does this occur ?
Very often, especially with changeset made with StreetComplete.

## Please tell us about your environment:
Firefox 95.0 on Fedora 35/x64_64.
Firefox 95.0 on Windows 10/x64_64.

**Comment by mmd-osm** [comment]
That’s very likely an overpass API limitation. StreetComplete is known for uploading multiple versions of an object in a very short amount of time. Overpass cannot distinguish those object versions, in case they were created at the exact same hour:minute :second, giving you incorrect results for a particular changeset.

**Comment by willemarcel** [comment]
As @mmd-osm said, it's an issue on Overpass and StreetComplete

**Comment by matkoniecz** [comment]
It is not a StreetComplete bug, as uploading multiple changesets shortly after each other is perfectly fine behaviour of OSM API.

Though it is anyway covered by #492

**Comment by mmd-osm** [comment]
Depends a bit on your point of view. Every other editing app would merge multiple changes to the same object into a single object version upload only. 

SC has its reasons to work differently. The more messages I see in my log files, the more I’m disagreeing with it.

**Comment by matkoniecz** [comment]
Hmm, adding one second delay may be actually doable. Open issue on SC issue tracker once you will be sufficiently disagreeable with it.

(no promising anything as I have no idea how this code works, but maybe it is possible to achieve more friendly behaviour)

**Comment by musuruan** [comment]
> That’s very likely an overpass API limitation. StreetComplete is known for uploading multiple versions of an object in a very short amount of time. Overpass cannot distinguish those object versions, in case they were created at the exact same hour:minute:second, giving you incorrect results for a particular changeset.
> 
> Related discussion: [streetcomplete/StreetComplete#2318 (comment)](https://github.com/streetcomplete/StreetComplete/issues/2318#issuecomment-868975126)

I do not agree. Achavi is also based on overpass but it doesn't show this behaviour.

For example, changeset #114816412:
https://nrenner.github.io/achavi/?changeset=114816412

**Comment by mmd-osm** [comment]
Let's take a look at the queries both tools are sending to the Overpass instances. First of all, you need to enable relations for Achavi, to have comparable queries: https://nrenner.github.io/achavi/?changeset=114816412&relations=true

Achavi: https://overpass-api.de/api/interpreter?data=[adiff:%222021-12-11T12:33:05Z%22,%222021-12-11T12:54:14Z%22];(node(bbox)(changed);way(bbox)(changed);relation(bbox)(changed););out%20meta%20geom(bbox);&bbox=8.0519203,45.5679148,8.052059,45.5685817

OSMCha: https://overpass.maptime.in/api/interpreter?data=[out:xml][adiff:%222021-12-11T12:32:58.000Z,%22,%222021-12-11T12:54:14Z%22];(node(bbox)(changed);way(bbox)(changed);relation(bbox)(changed););out%20meta%20geom(bbox);&bbox=8.0519203,45.5679148,8.052059,45.5685817

It seems that OSMCha is using an earlier start time: 12:32:58 vs. 12:33:05 in Achavi. Apart from this difference, I don't see other differences? Maybe this is reason for different results.

**Comment by nrenner** [comment]
This is a case of changesets overlapping by time and bbox [1] (shortened):
```
c114816420 s2021-12-11T12:33:23Z e2021-12-11T12:54:14Z TStreetComplete:quest_type=AddBuildingLevels,
c114816412 s2021-12-11T12:33:06Z e2021-12-11T12:54:14Z TStreetComplete:quest_type=AddWayLit,        
c114816402 s2021-12-11T12:32:59Z e2021-12-11T12:54:14Z TStreetComplete:quest_type=AddRoadSurface,   
```
StreetComplete opens multiple changesets in parallel, one for each active quest.

Each changeset has only one modify edit, two editing the same object [2] (shortened):
```
  <way id="340649056" version="3" changeset="114816420" timestamp="2021-12-11T12:33:23Z"
  <way id="314892545" version="6" changeset="114816412" timestamp="2021-12-11T12:33:07Z"
  <way id="314892545" version="5" changeset="114816402" timestamp="2021-12-11T12:32:59Z"
```

Using the times and bbox of changeset 114816402 ([metadata](https://www.openstreetmap.org/api/0.6/changeset/114816402), [OSMCha](https://osmcha.org/changesets/114816402?filters=%7B%22ids%22%3A%5B%7B%22label%22%3A%22114816402%2C114816412%2C114816420%22%2C%22value%22%3A%22114816402%2C114816412%2C114816420%22%7D%5D%2C%22date__gte%22%3A%5B%7B%22label%22%3A%22%22%2C%22value%22%3A%22%22%7D%5D%7D)) for the Overpass adiff query, it returns changes from other changesets that are within the same time range and bbox:
```
<way id="314892545" version="6" timestamp="2021-12-11T12:33:07Z" changeset="114816412"
<way id="340649056" version="3" timestamp="2021-12-11T12:33:23Z" changeset="114816420"
```
[https://overpass.maptime.in/api/interpreter?data=[out:xml][adiff:"2021-12-11T12:32:58.000Z,","2021-12-11T12:54:14Z"];(node(bbox)(changed);way(bbox)(changed);relation(bbox)(changed););out meta geom(bbox);&bbox=8.0519203,45.5679148,8.052059,45.5685817](https://overpass.maptime.in/api/interpreter?data=[out:xml][adiff:%222021-12-11T12:32:58.000Z,%22,%222021-12-11T12:54:14Z%22];(node(bbox)(changed);way(bbox)(changed);relation(bbox)(changed););out%20meta%20geom(bbox);&bbox=8.0519203,45.5679148,8.052059,45.5685817)

As the adiff query only returns the state at the given start and end time, not changes in between, version 5 of way 314892545 for this changeset is missing because of the later change in version 6.

### Other examples

by same user later in December (with filter listing all other changesets in edit session) [3]:

- [115164251](https://osmcha.org/changesets/115164251?filters=%7B%22date__gte%22%3A%5B%7B%22label%22%3A%22%22%2C%22value%22%3A%22%22%7D%5D%2C%22ids%22%3A%5B%7B%22label%22%3A%22115164251%2C115164268%22%2C%22value%22%3A%22115164251%2C115164268%22%7D%5D%7D)
- [115015641](https://osmcha.org/changesets/115015641?filters=%7B%22date__gte%22%3A%5B%7B%22label%22%3A%22%22%2C%22value%22%3A%22%22%7D%5D%2C%22ids%22%3A%5B%7B%22label%22%3A%22115015701%2C115015664%2C115015641%2C115015640%2C115015639%2C115015638%2C115015637%22%2C%22value%22%3A%22115015701%2C115015664%2C115015641%2C115015640%2C115015639%2C115015638%2C115015637%22%7D%5D%7D) ([S3 json](https://s3.amazonaws.com/mapbox/real-changesets/production/115015641.json) `elements` not empty as in other examples (?), changes from changeset 115015640),  
  [115015639](https://osmcha.org/changesets/115015639?filters=%7B%22date__gte%22%3A%5B%7B%22label%22%3A%22%22%2C%22value%22%3A%22%22%7D%5D%2C%22ids%22%3A%5B%7B%22label%22%3A%22115015701%2C115015664%2C115015641%2C115015640%2C115015639%2C115015638%2C115015637%22%2C%22value%22%3A%22115015701%2C115015664%2C115015641%2C115015640%2C115015639%2C115015638%2C115015637%22%7D%5D%7D),  
  [115015637](https://osmcha.org/changesets/115015637?filters=%7B%22date__gte%22%3A%5B%7B%22label%22%3A%22%22%2C%22value%22%3A%22%22%7D%5D%2C%22ids%22%3A%5B%7B%22label%22%3A%22115015701%2C115015664%2C115015641%2C115015640%2C115015639%2C115015638%2C115015637%22%2C%22value%22%3A%22115015701%2C115015664%2C115015641%2C115015640%2C115015639%2C115015638%2C115015637%22%7D%5D%7D) ([node 3202699546](https://osm.virtuelle-loipe.de/history/?type=node&ref=3202699546) v3 and v4 in same second)
- [114920616](https://osmcha.org/changesets/114920616?filters=%7B%22date__gte%22%3A%5B%7B%22label%22%3A%22%22%2C%22value%22%3A%22%22%7D%5D%2C%22ids%22%3A%5B%7B%22label%22%3A%22114920660%2C114920619%2C114920616%22%2C%22value%22%3A%22114920660%2C114920619%2C114920616%22%7D%5D%7D) 
- [114880955](https://osmcha.org/changesets/114880955?filters=%7B%22date__gte%22%3A%5B%7B%22label%22%3A%22%22%2C%22value%22%3A%22%22%7D%5D%2C%22ids%22%3A%5B%7B%22label%22%3A%22114880962%2C114880955%2C114880952%22%2C%22value%22%3A%22114880962%2C114880955%2C114880952%22%7D%5D%7D)



### What OSMCha could do

1. Filter out other changesets from the fallback query result.  
   I suppose this is already done in the server (see also [changeset-map#103](https://github.com/osmlab/changeset-map/pull/103/files)), therefore the `elements` property is empty in the [S3 json](https://s3.amazonaws.com/mapbox/real-changesets/production/114816402.json).
2. Don't run the client-side fallback query when the server-side was successful and an [empty result](https://github.com/osmlab/changeset-map/blob/a7fc61c8cf1aa362e728e20cd819cf03270603fc/lib/getChangeset.js#L16) is expected with the current implementation
3. Request missing versions with an additional query, e.g. using [timeline](https://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_QL#The_statement_timeline)/[retro](https://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_QL#The_block_statement_retro) ([blog](https://dev.overpass-api.de/blog/sliced_time_and_space.html#timeline)):
   ```
   [out:json];
   timeline(way, 314892545, 5);
   for (t["created"])
   {
     retro (_.val)
     {
       way(314892545);
       out meta geom;
     }
   }
   ```
   [Overpass Turbo](http://overpass-turbo.eu/?Q=%0A%5Bout%3Ajson%5D%3B%0Atimeline(way%2C%20314892545%2C%205)%3B%0Afor%20(t%5B%22created%22%5D)%0A%7B%0A%20%20retro%20(_.val)%0A%20%20%7B%0A%20%20%20%20way(314892545)%3B%0A%20%20%20%20out%20meta%20geom%3B%0A%20%20%7D%0A%7D&C=45.56825;8.05199;18&R)  
   Limitation: does not work for same-second versions  (like [node 3202699546](https://osm.virtuelle-loipe.de/history/?type=node&ref=3202699546) v3, [Overpass Turbo](http://overpass-turbo.eu/?Q=%2F%2F%20misses%20v3%20(same%20second%20as%20v4)%0A%5Bout%3Ajson%5D%3B%0Atimeline(node%2C%203202699546)%3B%0Aout%3B%0A&C=45.5588;8.0543;18&R))

----

### Footnotes

- [1] `curl -s "https://www.openstreetmap.org/api/0.6/changesets?display_name=xxx&time=2021-12-11T12:32:59Z,2021-12-11T12:54:14Z" | osmium cat -F .osm.xml -f .opl`  
  changesets API call using [changeset 114816402 metadata](https://www.openstreetmap.org/api/0.6/changeset/114816402) user (anonymized), created_at and closed_at, formatted as osmium-tool OPL (object per line)
- [2] `curl -s https://www.openstreetmap.org/api/0.6/changeset/{114816420,114816412,114816402}/download | grep changeset` 
- [3] get later changesets by user (anonymized), manually group into edit sessions (same/close changeset closed timestamp), check each session for empty json `elements` using `Content-Length` HTTP header
  - `wget -qO - "https://www.openstreetmap.org/api/0.6/changesets?display_name=xxx&time=2021-12-11T12:32:59Z" | osmium cat -F .osm.xml -f .opl` 
  - `curl -s --head https://s3.amazonaws.com/mapbox/real-changesets/production/{114816402,114816412,114816420}.json | grep "Content-Length"`
